### PR TITLE
refactor type system (a rough idea)

### PR DESCRIPTION
**Pull request by ianhe8x**
related to #843 and #580 
Here's a rough idea of the refactor.
Goals:
1. decouple the code from a specified Metadata version.
2. change global injection to api scope
3. [Optional] type name resolve improvement, resolve duplicate name across runtimes properly.

Approach
For Goal 1):
1. change type using classes from MetadataV0 to interface.
2. types-storage and types-extrinsic shall handle Metadata of different version with specified logic.
For Goal 2):
1. whereever call createType or global Injection now call typeRegistry. 
2. all places `new Type()` shall be wrapped by `TypeRegistry.withTypeRegistry()` then inside can use TypeRegistry's static property to get access to the api-scope TypeRegistry instance.
For Goal 3):
1. after all type resolving are restrained in TypeRegistry, we should be able to come up with a solution.

### CODE REVIEW

File: packages/types/src/type/Event.ts [hunk 1]
```diff
@@ -86,18 +85,18 @@ export class EventIndex extends U8aFixed {
 export default class Event extends Struct {
   // Currently we _only_ decode from Uint8Array, since we expect it to
   // be used via EventRecord
-  constructor (_value?: Uint8Array) {
-    const { DataType, value } = Event.decodeEvent(_value);
+  constructor (_value: Uint8Array = new Uint8Array(), typeRegistry: TypeRegistry) {
```
Comment (amaurymartiny) [inline packages/types/src/type/Event.ts]: Yes, exactly the idea Jaco and I had some time ago: dependency injection via constructor. Teh tricky part is to figure out how to pass in `typeRegistry` to this constructor.

If i'm not mistaken, we instantiate this class here: https://github.com/polkadot-js/api/blob/master/packages/rpc-core/src/index.ts#L182. Which means that we somehow need to pass the typeRegistry to rpc-core too. I believe it's ultimately the correct way to go. We basically need to have `typeRegistry` be passed around in a lot of function arguments.

**Comment by amaurymartiny** [comment]
I think Goal 1 can be done in a separate PR, and afaict this PR is addressing Goal 2

**Comment by ianhe8x** [comment]
at first the idea is just pass typeRegistry in constructor, but after starting it, I feel this approach not very ideal regarding
1. before we assume the first arg of constructor is always `value`, now we might have to add another that second arg is typeRegistry, even in where typeRegistry is not be needed.
2. it also means we will have to pass typeRegistry through a deep call stack to where it's used.

The use of static holder of typeRegistry is a tricky but effective way to avoid a large scale of changing in signature.

**Comment by amaurymartiny** [comment]
> The use of static holder of typeRegistry is a tricky but effective way to avoid a large scale of changing in signature.

How it this different than the current solution of a singleton? It just _seems_ more "organized" since it's in a class, but it's still a singleton

**Comment by ianhe8x** [comment]
it is api scoped, check https://github.com/polkadot-js/api/blob/b0270056242b089aac9c37bc247c87377ff9b811/packages/type-extrinsics/src/fromMetadata.spec.ts#L12
as long as the entry use withTypeRegistry and providing a instance of registry, the static one will be replaced by that with the call lifecycle, thanks for the fact js is single-threaded, we don't need to worry about thread safety.

**Comment by ianhe8x** [comment]
in Struct.ts I put a protected typeRegistry member, so after an instance is created, if it has further need to access to typeregister, it can use the instance one since the static one will not be available at that moment.

**Comment by ianhe8x** [comment]
Finally get all e2e tests working on my local node. I will start fixing other unit tests and clean the codes. Need advice @amaurymartiny @jacogr .

File: packages/types/src/codec/TypeRegistry.ts [hunk 2]
```diff
@@ -0,0 +1,319 @@
...
+  static get TYPE_REGISTRY (): TypeRegistry {
+    assert(TypeRegistry._TYPE_REGISTRY, 'TYPE_REGISTRY is undefined, make sure it is wrapped in withRegistry()');
+    return TypeRegistry._TYPE_REGISTRY as TypeRegistry;
+  }
+
+  static withRegistry (typeRegistry: TypeRegistry, wrapFn: (...args: any[]) => any) {
+    TypeRegistry.TY_STACK.push(TypeRegistry._TYPE_REGISTRY);
+    TypeRegistry._TYPE_REGISTRY = typeRegistry;
+    const ret = wrapFn();
+    if (isPromise(ret)) {
```
Comment (ianhe8x) [inline packages/types/src/codec/TypeRegistry.ts]: Supporting promise is probably a bad idea, the stack could be messed if two promises are resolved in a wrong order. will remove promise support later.

**Comment by amaurymartiny** [comment]
If I understood you correctly:

`withRegistry(api.typeRegistry, () => f(arg1, arg2))` will:
  1. save current value of `TypeRegistry._TYPE_REGISTRY` to the stack
  2. temporarily set the singleton `TypeRegistry._TYPE_REGISTRY` to `api.typeRegistry`
  3. `f` will be called, and `f` can access the singleton `TypeRegistry.TYPE_REGISTRY` in its execution lifetime
  4. Once `f` is finished, we pop `TypeRegistry.TYPE_REGISTRY` to the value before

TBH I find this **really** hacky.

I would propose to see the following:

```diff
- withRegistry(api.typeRegistry, () => f(arg1, arg2))
+ withRegistry(api.typeRegistry, (registry) => f(arg1, arg2, registry))
```
and we need to refactor everywhere that uses the singleton. Reason: pure functions, easier to maintain, dependencies are clearly stated.

**Comment by ianhe8x** [comment]
Well, if we pass typeRegistry all the way down to where it's needed, there's no need for withRegistry().
I tried and am really not convinced that passing it through is a better idea than this one.
Another way might work, which is instead of `export class `, we export `(typeRegistry)=>class xxx`. so we can still keep constructor as before, in additionally have an api scope typeRegistry access. But a large scale of signature changes are also inevitable.

**Comment by amaurymartiny** [comment]
> Well, if we pass typeRegistry all the way down to where it's needed, there's no need for withRegistry().

> But a large scale of signature changes are also inevitable.

Agreed with both. TBH thinking long-term, I think that's the cleanest way to go.

**Comment by ianhe8x** [comment]
To do that, we have to make it a convention that second parameter in all Codec's constructor is typeregister, even for types who don't need it. Otherwise it would be hard for Tuple, Vector and othet boxing types to decide whether and how to pass typeregister. And then for types don't have value as first parameter we need to use it like new Type(undefined, typeregister). Feels terrible and could cause confusion

**Comment by jacogr** [comment]
There is a _LOT_ of work here. My 2 satoshis -

I'm certainly very cognisant on the ripple effects elsewhere, that is not inside our control, i.e. https://github.com/hicommonwealth/edgeware-node-types If it was all internally, I would be much more confident instead of us forcing updates everywhere (which is dead work is we ever get to the elusive #1328)

Make no mistake, we need to break the global link and have it local to API, but at the same time I feel we are solving a something that we can solve in cleaner ways down the line if we didn't have any external types being injected.

**Pull request closed without merging**